deterministic-instantsend.md

```diff
@@ -0,0 +1,112 @@
+<pre>
+DIP: 00xx
+Title: Making InstantSend Deterministic with use of Quorum Cycles
+Author: Samuel Westrich, UdjinM6
+Special-Thanks: Thephez, Virgile Bartolo
+Comments: No comments yet.
+Status: Draft
+Layer: Consensus (hard fork)
+Created: 2020-08-02
+License: MIT License
+</pre>
+
+## Abstract
+
+This DIP aims to improve InstantSend messages to make them deterministically verifiable.
+
+## Motivation
+
+LLMQ based InstantSend was introduced in DIP10. In that implementation InstantSend locks are only
+verifiable by recent quorums because the InstantSend lock does not include any block or time based
+information. In Dash Platform InstantSend Locks are used to add credit to Identities as they provide
+input finality. When blocks are replayed on the Platform Chain, all State Transitions need to be
+re-validated and it is possible that InstantSend signatures will need to be rechecked. However, to
+recheck them one needs to know the quorum that signed them. In this DIP we will provide a mechanism
+to that end.
+
+## Previous work
+
+* [DIP-0006: Long-Living Masternode Quorums](https://github.com/dashpay/dips/blob/master/dip-0006.md)
+* [DIP-0007: LLMQ Signing Requests / Sessions](https://github.com/dashpay/dips/blob/master/dip-0007.md)
+* [DIP-0010: LLMQ InstantSend](https://github.com/dashpay/dips/blob/master/dip-0010.md)
+
+## Versioning of ISLock messages
+
+Since `islock` messages were never versioned, a new `ISDLOCK` message will be created and the
+`islock` message will be deprecated. ISD stands for InstantSend Deterministic. The version of the
+`ISDLOCK` used in this document will be `1`. We will still refer to `islock` messages, even though
+the message name has been changed.
+
+The `messageHash` for `islock` messages should now be calculated as `SHA256(version, txHash).`
+
+## QuorumHash vs CycleHash
+
+The naive approach to fixing this problem would be to include the `quorumHash` in the `islock`
+message. An `islock` would then be easily verifiable since the quorum that signed it would always be
+known. The drawback of this approach is that any quorum could sign any `islock` even if it were not
+responsible to do so for that quorum cycle.
+
+A quorum cycle begins at a `quorumBlock` (as per
+[DIP-0006](https://github.com/dashpay/dips/blob/master/dip-0006.md#parametersvariables-of-a-llmq-and-dkg))
+and lasts for a number of blocks that is equal to the `quorumDkgInterval`. During this time the set
+of valid quorums are not modified for a given `quorumType`. If the `quorumDkgInterval` is set to 24
+blocks then from block 100 to block 124 quorums of that type stay the same. During this period the
+quorum that is in charge of signing a specific `islock` will also always be the same. CycleHash is
+the `blockHash` of the first block in a cycle.
```

Comment:
Is this the agreed upon definition for cycle hash? There was previously some discussion regarding exactly which block's hash should be used. @QuantumExplorer @UdjinM6